crates/ruff_db/src/file_system/vendored.rs

```diff
@@ -0,0 +1,353 @@
+        let mut buffer = String::with_capacity(zipfile.size() as usize);
+        zipfile.read_to_string(&mut buffer)?;
+        Ok(buffer)
+    }
+}
+
+impl FileSystem for VendoredFileSystem {
+    fn exists(&self, path: &FileSystemPath) -> bool {
+        let normalized = normalize_vendored_path(path);
+        self.vendored_archive().lookup_path(&normalized).is_ok()
+            || (!normalized.has_trailing_slash()
+                && self
+                    .vendored_archive()
+                    .lookup_path(&normalized.with_trailing_slash())
+                    .is_ok())
+    }
+
+    fn is_directory(&self, path: &FileSystemPath) -> bool {
+        let normalized = normalize_vendored_path(path);
+        if let Ok(zipfile) = self.vendored_archive().lookup_path(&normalized) {
+            return zipfile.is_dir();
+        }
+        if normalized.has_trailing_slash() {
+            return false;
+        }
+        self.vendored_archive()
+            .lookup_path(&normalized.with_trailing_slash())
+            .is_ok_and(|zipfile| zipfile.is_dir())
+    }
+
+    fn is_file(&self, path: &FileSystemPath) -> bool {
+        if path.as_str().ends_with('/') {
+            return false;
+        }
+        self.vendored_archive()
+            .lookup_path(&normalize_vendored_path(path))
+            .is_ok_and(|zipfile| zipfile.is_file())
+    }
+
+    fn metadata(&self, path: &FileSystemPath) -> Result<Metadata> {
+        let normalized = normalize_vendored_path(path);
+        let mut archive = self.vendored_archive();
+
+        let is_dir = match archive.lookup_path(&normalized) {
+            Ok(zipfile) => zipfile.is_dir(),
+            Err(err) => {
+                if normalized.has_trailing_slash() {
+                    return Err(err);
+                }
+                let mut archive = self.vendored_archive();
+                let lookup_result = archive.lookup_path(&normalized.with_trailing_slash());
+                if let Ok(zipfile) = lookup_result {
+                    zipfile.is_dir()
+                } else {
+                    return Err(err);
+                }
+            }
+        };
+
+        let file_type = if is_dir {
+            FileType::Directory
+        } else {
+            FileType::File
+        };
+
+        Ok(Metadata {
+            revision: FileRevision::zero(),
+            permissions: Some(Self::READ_ONLY_PERMISSIONS),
+            file_type,
+        })
+    }
+
+    fn read(&self, path: &FileSystemPath) -> Result<String> {
+        let normalized = normalize_vendored_path(path);
+        let mut archive = self.vendored_archive();
+        let lookup_error = match archive.lookup_path(&normalized) {
+            Ok(zipfile) => return Self::read_zipfile(zipfile),
+            Err(err) => err,
+        };
+        if normalized.has_trailing_slash() {
+            return Err(lookup_error);
+        }
+        let mut archive = self.vendored_archive();
```

Comment:
Can't we re-use `archive` here?
```suggestion
```

Reply:
No, or the borrow checker complains that we're mutably borrowing it twice in the same scope.

Reply:
This seems to compile fine

```rust
let normalized = normalize_vendored_path(path);
        let mut archive = self.vendored_archive();
        let lookup_error = match archive.lookup_path(&normalized) {
            Ok(zipfile) => return Self::read_zipfile(zipfile),
            Err(err) => err,
        };
        if normalized.has_trailing_slash() {
            return Err(lookup_error);
        }
        // let mut archive = self.vendored_archive();
        let zipfile = archive.lookup_path(&normalized.with_trailing_slash())?;
        Self::read_zipfile(zipfile)
```

Reply:
So it does. I'm sure I tried that yesterday. Ah well -- thanks!